### ISSUE TITLE
[jss-isolate][jss-nested] jss-isolate isolates nested rules

### ISSUE THREAD

**Issue by yakhinvadim** [cssinjs/jss#698]
Consider the code: 
```js
const styles = {
    link: {
        color: 'green',
        '&:hover': {
            textDecoration: 'underline'
        }
    } 
}
```

With jss-isolate, I'd expect the following behaviour:
```css
.link-123456 {
    color: green;
}

/* isolated */
.link-123456:hover {
    /* color is inherited from the rule above */
    text-decoration: underline;
}
```

but got:
```css
.link-123456 {
    color: green;
}

/* isolated */
.link-123456:hover {
    /* color is set to black by isolation */
    text-decoration: underline;
}
```

According to this comment, it's not the expected behaviour: https://github.com/cssinjs/jss/issues/334#issuecomment-263887054

I've found three ways to workaround this:
1) add `isolate: false`
    ```js
    const styles = {
        link: {
            color: 'green',
            '&:hover': {
                isolate: false,
                textDecoration: 'underline'
            }
        } 
    }
    ```
    This exposes the rule to external styles, so 😕

2) duplicate rules
    ```js
    const styles = {
        link: {
            color: 'green',
            '&:hover': {
                color: 'green',
                textDecoration: 'underline'
            }
        } 
    }
    ```
    This may lead to the situation, where developers forget to change both rules, so 😕 too

3) extract rules to variable
    ```js
    const linkStyles = {
        color: 'green',
    }

    const styles = {
        link: {
            ...linkStyles,
            '&:hover': {
                ...linkStyles,
                textDecoration: 'underline'
            }
        } 
    }
    ```
    This is the best so far, but too verbose.

Is there any other way to get the expected behaviour?

**Comment by kof** [comment]
jss-isolate has currently no knowledge about nested rules, I used the
tactic of setting isolate: false for nested rules.

Using root convention and only isolate those also might be a good option if
you don't have to integrate in a hostile environment.

It is hard to avoid double isolation for nested rules. I was thinking to
create a separate plugin for pseudo selectors and teach jss-isolate to not
double-isolate those.

**Comment by yakhinvadim** [comment]
Thanks! I guess, we'll use "isolate: false" approach for now, because our environments are not that hostile.

I was also thinking about using jss-compose, but it doesn't support nested rules.
Then, I was trying to use jss-extend like this:
```js
const styles = {
    link: {
        color: 'green',
        '&:hover': {
            extend: 'link',
            textDecoration: 'underline'
        }
    } 
}
```

But that led to infinite recursion. What do you think about adding an exception for jss-extend, so calling it from nested rule won't cause infinite recursion? 

I'm afraid a separate plugin for pseudo selectors will confuse users, because it kind of does the same thing as jss-nested with a little different syntax.

**Comment by kof** [comment]
Also in this case you don't have to set isolate: false for pseuodo selectors

**Comment by yakhinvadim** [comment]
I have a little different question, but it's highly related, so I'll ask it here.

Isolation by convention is great: we protect our components from all inherited styles, but still, allow inheritance inside components.

**But what about non-inherited styles set globally?**

Let's say we develop a simple component with `<h1>` in it. We don't set margin explicitly, assuming it will be browser's default: `margin-top: 0.67em; margin-bottom: 0.67em`.
But our component is used in app, where bootstrap styles are present:
```css
h1, h2, h3, h4, h5, h6 {
  margin-top: 0;
  margin-bottom: 0.5rem;
}
```
(see https://github.com/twbs/bootstrap/blob/v4-dev/dist/css/bootstrap-reboot.css#L52)

So, in this app our components will end up with broken styles. 

It would be great if we could use two types of isolation together:
* For 'root' classes - isolation by convention of **all properties**
* For other classes - isolation of **all non-inherited properties**

I've tried to come up with an option object for jss-isolate, which would add this behavior, but as I understand, it is now impossible to combine these two types.

So, my questions are: 
1) Is it possible? And if not:
2) Do you think such setup will be useful?

**Comment by kof** [comment]
> But what about non-inherited styles set globally?

thats why I asked if you control the environment, if you do - don't use global declarations

**Comment by kof** [comment]
also see https://github.com/cssinjs/reset-jss and https://github.com/cssinjs/normalize-jss

**Comment by kof** [comment]
> Do you think such setup will be useful?

yeah I think this could be useful, currently it won't work, it would require a way to define this

**Comment by yakhinvadim** [comment]
> yeah I think this could be useful, currently it won't work, it would require a way to define this

Good, then if you don't mind, I'll try to make a PR for this in a couple of weeks.

**Comment by kof** [comment]
Sure!

**Comment by kof** [comment]
Lets create a new issue with a spec?

**Comment by yakhinvadim** [comment]
Sure, I've made a new issue for that. Right now it's kind of a placeholder, so I'll make it more complete later. Actually, I might not fully understand what is needed in the spec, so feel free to correct me :)

**Comment by DiesIrae** [comment]
We just bumped on this issue in my organization. We are putting a widget in a hostile environment, this is why we would like to use jss-isolate in the first place, not to be polluted by global changes. All solutions exposed by @yakhinvadim are impractical for us.

Making compatible jss-isolate and jss-nested is still not a priority? Is this a strong technical challenge, or is this plugin simply not evolving any more (I see the last code change dates from 2 years ago)?

Thanks!
